server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java

```diff
@@ -19,48 +19,313 @@
+            NOT_EXACTLY_EQUALS("not-exactly-equals");
+            
+            public static Optional<Comparator> find(String comparatorName) {
+                return Arrays.stream(values())
+                        .filter(value -> value.comparatorName.equalsIgnoreCase(comparatorName))
+                        .findAny();
+            }
+            
+            public static Comparator of(String comparatorName) {
+                return find(comparatorName).orElseThrow(() -> new IllegalStateException("'" + comparatorName + "' is not a valid comparator name"));
+            }
+            
+            private final String comparatorName;
+            
+            private Comparator(String comparator) {
+                this.comparatorName = comparator;
+            }
+            
+            public String asString() {
+                return comparatorName;
+            }
+        }
+
+        public static Condition of(String field, String comparator, String value) {
+            Preconditions.checkNotNull(field, "field should no be null");
+            Preconditions.checkNotNull(comparator, "comparator should no be null");
+            Preconditions.checkNotNull(value, "value should no be null");
+            Preconditions.checkArgument(StringUtils.isNotBlank(value), "value should no be empty");
+            return new Condition(Field.of(field), Comparator.of(comparator), value);
+        }
+
+        private final Field field;
+        private final Comparator comparator;
+        private final String value;
+
+        private Condition(Field field, Comparator comparator, String value) {
+            this.field = field;
+            this.comparator = comparator;
+            this.value = value;
+        }
+        
+        public Field getField() {
+            return field;
+        }
+        
+        public Comparator getComparator() {
+            return comparator;
+        }
+        
+        public String getValue() {
+            return value;
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof Condition) {
+                Condition condition = (Condition) o;
+                return Objects.equals(field, condition.field)
+                    && Objects.equals(comparator, condition.comparator)
+                    && Objects.equals(value, condition.value);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(field, comparator, value);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                .add("field", field)
+                .add("comparator", comparator)
+                .add("value", value)
+                .toString();
+        }
+    }
+
+
+    public static class Action {
+
+        public static Action ofMailboxIds(List<String> mailboxIds) {
```

Comment:
if you don't want to enforce ImmutableList on input, you should copy it before keeping a reference to it